src/backend/distributed/planner/multi_logical_optimizer.c

```diff
@@ -1959,6 +2021,145 @@ WorkerExtendedOpNode(MultiExtendedOp *originalOpNode,
+ * has non-partition column reference in aggregate definition. Note that, it only
+ * checks aggs subfield of Aggref, it does not check FILTER or SORT clauses.
+ */
+static bool
+HasNonPartitionColumnDistinctAgg(List *targetEntryList, Node *havingQual,
+								 List *tableNodeList)
+{
+	List *targetVarList = pull_var_clause((Node *) targetEntryList,
+										  PVC_INCLUDE_AGGREGATES);
+	List *havingVarList = pull_var_clause((Node *) havingQual, PVC_INCLUDE_AGGREGATES);
+	List *aggregateCheckList = list_concat(targetVarList, havingVarList);
+	ListCell *aggregateCheckCell = NULL;
+	foreach(aggregateCheckCell, aggregateCheckList)
+	{
+		Node *targetNode = lfirst(aggregateCheckCell);
+		Aggref *targetAgg = NULL;
+		List *varList = NIL;
+		ListCell *varCell = NULL;
+		bool isPartitionColumn = false;
+
+		if (IsA(targetNode, Var))
+		{
+			continue;
+		}
+
+		Assert(IsA(targetNode, Aggref));
+		targetAgg = (Aggref *) targetNode;
+		if (targetAgg->aggdistinct == NIL)
+		{
+			continue;
+		}
+
+		varList = pull_var_clause_default((Node *) targetAgg->args);
+
+		if (varList == NIL)
+		{
+			continue;
+		}
+
+		foreach(varCell, varList)
+		{
+			Node *targetVar = (Node *) lfirst(varCell);
+			Assert(IsA(targetVar, Var));
+
+			isPartitionColumn = ExpressionContainsPartitionColumn(
+				targetVar, tableNodeList);
+
+			if (!isPartitionColumn)
+			{
+				return true;
+			}
+		}
+	}
+
+	return false;
+}
+
+
+/*
+ * ShouldPullDistinctColumn returns true if distinct aggregate should pull
+ * individual columns from worker to master and evaluate aggregate operation
+ * at master.
+ *
+ * Pull cases are:
+ * - repartition subqueries
+ * - query has count distinct on a non-partition column on at least one target
+ * - count distinct is on a non-partition column and query is not
+ *   grouped on partition column
+ */
+static bool
+ShouldPullDistinctColumn(bool groupedByDisjointPartitionColumn,
+						 bool repartitionSubquery,
+						 bool hasNonPartitionColumnDistinctAgg,
+						 Node *expression, List *tableNodeList)
+{
+	bool hasPartitionColumn = false;
+
+	if (!hasNonPartitionColumnDistinctAgg)
+	{
+		hasPartitionColumn = ExpressionContainsPartitionColumn(expression,
+															   tableNodeList);
+	}
+	if (repartitionSubquery)
+	{
+		return true;
+	}
+	else if (!groupedByDisjointPartitionColumn && !hasPartitionColumn)
+	{
+		return true;
+	}
+
+	return false;
+}
+
+
+/*
+ * ExpressionContainsPartitionColumn returns true if provided expression
+ * contains partition column reference from provided table node list.
+ * It also returns false if a column is partition column of an append
+ * distributed table.
+ */
+static bool
+ExpressionContainsPartitionColumn(Node *expression, List *tableNodeList)
+{
+	List *targetColumnList = pull_var_clause_default((Node *) expression);
+	Var *firstColumn = NULL;
+	ListCell *tableNodeCell = NULL;
+
+	if (targetColumnList == NIL)
+	{
+		return false;
+	}
+
+	firstColumn = (Var *) linitial(targetColumnList);
```

Comment:
Can `targetColumnList` have more than one colum(Var)?

Reply:
It should not. We error out if count distinct has more than one column.

Reply:
then maybe we can rename `firstColumn` to `distinctColumn`

Reply:
sometimes it can :(
when there is a non-distinct aggregate

Reply:
It should be more like, ```HasPartitionColumnOnly```

Reply:
For our purpose : we are only interested in value of this function if a target expression is count distinct. In this case count(distinct ..) can except only columns. So we should be fine as soon as the function is called within this context.

Reply:
updated the function comment. kept the ```firstColumn``` name the same